presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -643,6 +647,101 @@ private void validateMaterialziedViewQueryPlan(Statement query)
             validator.process(query, new MaterializedViewPlanValidatorContext());
         }
 
+        @Override
+        protected Scope visitRefreshMaterializedView(RefreshMaterializedView node, Optional<Scope> scope)
+        {
+            analysis.setUpdateType("INSERT");
+
+            QualifiedObjectName viewName = createQualifiedObjectName(session, node.getTarget(), node.getTarget().getName());
+
+            ConnectorMaterializedViewDefinition view = metadata.getMaterializedView(session, viewName)
+                    .orElseThrow(() -> new SemanticException(MISSING_MATERIALIZED_VIEW, node, "Materialized view '%s' does not exist", viewName));
+
+            accessControl.checkCanInsertIntoTable(session.getRequiredTransactionId(), session.getIdentity(), session.getAccessControlContext(), viewName);
+
+            // Analyzer will check for SELECT permission on materialized view table, but REFERSH MATERIALIZED VIEW only needs INSERT permission.
+            // So use a new StatementAnalyzer instance with access check disabled to analyze the materialized view table.
+            StatementAnalyzer analyzer = new StatementAnalyzer(
+                    analysis,
+                    metadata,
+                    sqlParser,
+                    new AllowAllAccessControl(),
+                    session,
+                    warningCollector);
+            Scope viewScope = analyzer.analyze(node.getTarget(), scope);
+
+            Map<SchemaTableName, Expression> tablePredicates = RefreshMaterializedViewPredicateAnalyzer.extractTablePredicates(viewName, node.getWhere(), viewScope, metadata, session);
+
+            Query viewQuery = parseView(view.getOriginalSql(), viewName, node);
+            Query refreshQuery = viewQuery;
+            if (tablePredicates.containsKey(toSchemaTableName(viewName))) {
+                refreshQuery = buildQueryWithPredicate(viewQuery, tablePredicates.get(toSchemaTableName(viewName)));
```

Comment:
I am very confused about this approach. I am not really sure why are we distributing result query building in to parts. One here and another in the visit table. 

In my option we should do all the query builder work at one place so that it is easy to understand and debug.

We should also leverage predicateStitcher to help with predicate stitching for base tables, as it would require building predicates for the base tables